core/Session/SessionFingerprint.php

```diff
@@ -0,0 +1,102 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+
+namespace Piwik\Session;
+
+use Piwik\Date;
+
+/**
+ * Manages session information that is used to identify who the session
+ * is for.
+ *
+ * Once a session is authenticated using either a user name & password or
+ * token auth, some information about the user is stored in the session.
+ * This info includes the user name and the user agent
+ * string of the user's client, and a random session secret.
+ *
+ * In subsequent requests that use this session, we use the above information
+ * to verify that the session is allowed to be used by the person sending the
+ * request.
+ *
+ * This is accomplished by checking the request's user agent
+ * against what is stored in the session. If it doesn't then this is a
+ * session hijacking attempt.
+ *
+ * We also check that a hash in the piwik_auth cookie matches the hash
+ * of the time the user last changed their password + the session secret.
+ * If they don't match, the password has been changed since this session
+ * started, and is no longer valid.
+ */
+class SessionFingerprint
+{
+    const USER_NAME_SESSION_VAR_NAME = 'user.name';
+    const SESSION_INFO_SESSION_VAR_NAME = 'session.info';
+
+    public function getUser()
+    {
+        if (isset($_SESSION[self::USER_NAME_SESSION_VAR_NAME])) {
+            return $_SESSION[self::USER_NAME_SESSION_VAR_NAME];
+        }
+
+        return null;
+    }
+
+    public function getUserInfo()
+    {
+        if (isset($_SESSION[self::SESSION_INFO_SESSION_VAR_NAME])) {
+            return $_SESSION[self::SESSION_INFO_SESSION_VAR_NAME];
+        }
+
+        return null;
+    }
+
+    public function initialize($userName, $time = null, $userAgent = null)
+    {
+        $_SESSION[self::USER_NAME_SESSION_VAR_NAME] = $userName;
+        $_SESSION[self::SESSION_INFO_SESSION_VAR_NAME] = [
```

Comment:
very random thought and very likely not an issue... AFAIK PHP uses some custom serialize format for sessions. Not sure if there are any known issues re security and whether users could potentially set any weird username (eg when `disable_checks_usernames_attributes` enabled) or user agent to somehow escape and make the result of the session a different username. Only trying to think a bit out of the box... I really don't think it is an issue but wanted to just mention it.

Was thinking to maybe sanitize username/useragent maybe but likely not needed and could maybe even have other security issues when doing that...

Reply:
We could maybe add an automated test w/ every symbol & some unicode characters (maybe w/ a null character in it) and see if the session user info is written/read correctly.

Reply:
could but not sure it helps much as even the combination of it may be relevant. Only wanted to mention it but not much of a concern AFAIK